tests/tests/test_portforward.py

```diff
@@ -0,0 +1,173 @@
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+import os
+import subprocess
+import time
+
+from multiprocessing import Process
+from tempfile import NamedTemporaryFile
+
+from DNS import DnsRequest
+
+from ..common_setup import standard_setup_one_client
+from ..MenderAPI import authentication, devauth, get_container_manager, logger
+from .common_connect import wait_for_connect
+from .common import md5sum
+from .mendertesting import MenderTesting
+
+
+class TestPortForward(MenderTesting):
+    """Tests the port forward functionality"""
+
+    def test_portforward(self, standard_setup_one_client):
+        # accept the device
+        devauth.accept_devices(1)
+
+        # list of devices
+        devices = list(
+            set([device["id"] for device in devauth.get_devices_status("accepted")])
+        )
+        assert 1 == len(devices)
+
+        # wait for the device to connect via websocket
+        auth = authentication.Authentication()
+        wait_for_connect(auth, devices[0])
+
+        # device ID and auth token
+        devid = devices[0]
+        assert devid != None
+
+        # authenticate with mender-cli
+        server_url = "https://" + get_container_manager().get_mender_gateway()
+        username = auth.username
+        password = auth.password
+        p = subprocess.Popen(
+            [
+                "mender-cli",
+                "--skip-verify",
+                "--server",
+                server_url,
+                "login",
+                "--username",
+                username,
+                "--password",
+                password,
+            ],
+            stdout=subprocess.PIPE,
+        )
+        stdout, stderr = p.communicate()
+        exit_code = p.wait()
+        assert exit_code == 0, (stdout, stderr)
+
+        # start the port forwarding session in a different thread
+        def port_forward(server_url, dev_id, port_mapping, *port_mappings):
+            p = subprocess.Popen(
+                [
+                    "mender-cli",
+                    "--skip-verify",
+                    "--server",
+                    server_url,
+                    "port-forward",
+                    devid,
+                    port_mapping,
+                ]
+                + list(port_mappings),
+                stdout=subprocess.PIPE,
+            )
+            stdout, stderr = p.communicate()
+            p.wait()
+
+        p = Process(
+            target=port_forward,
+            args=(server_url, devid, "8822:22", "udp/8853:8.8.8.8:53"),
```

Comment:
Just to reduce the risk of confusion: Could we pick a different port than 8822? This port is used as the default ssh port into QEMU instances; using a different port number would make it clearer that you are not taking that route here (connecting to `localhost`, not the QEMU IP).